workflow/controller/pod_cleanup.go

```diff
@@ -1,22 +1,142 @@
+// we should add a common.LabelKeyCompleted=true label
+func getLabelSelector(wfName string, podGC *wfv1.PodGC, labelCompleted bool) metav1.LabelSelector {
+	// upper limit on memory usage set
+	labels := make(map[string]string)
+	labels[common.LabelKeyWorkflow] = wfName
+	if labelCompleted {
+		labels[common.LabelKeyCompleted] = "true"
+	}
+
+	expressions := []metav1.LabelSelectorRequirement{}
+
+	if podGC != nil && podGC.LabelSelector != nil {
+		for key, value := range podGC.LabelSelector.MatchLabels {
+			labels[key] = value
+		}
+		expressions = append(expressions, podGC.LabelSelector.MatchExpressions...)
+	}
+	return metav1.LabelSelector{MatchLabels: labels, MatchExpressions: expressions}
+}
+
+func (woc *wfOperationCtx) runImmediateCleanup(pod *apiv1.Pod, podGC *wfv1.PodGC, workflowPhase wfv1.WorkflowPhase) error {
+	/// we have finished the cleanup previously
+	if woc.finishedCleanup {
+		return nil
+	}
+
+	log := woc.log.WithField("podName", pod.Name)
+	log.Infof("Running PodGC strategy PodGCOnWorkflowCompletion for %s", pod.Name)
+	pods := woc.controller.kubeclientset.CoreV1().Pods(pod.Namespace)
+	wfName := woc.execWf.ObjectMeta.Name
+	podGCSelector, err := podGC.GetLabelSelector()
+	if err != nil {
+		return err
+	}
+	woc.rateLimiter.Wait()
+	// check if we need to label the pod as completed first and do so if we must
+	if determinePodCleanupAction(podGCSelector, pod.Labels, wfv1.PodGCOnPodCompletion, workflowPhase, pod.Status.Phase) == labelPodCompleted {
+		_, err := pods.Patch(
+			context.Background(),
+			pod.Name,
+			types.MergePatchType,
+			[]byte(`{"metadata": {"labels": {"workflows.argoproj.io/completed": "true"}}}`),
+			metav1.PatchOptions{},
+		)
+		if err != nil {
+			log.Errorf("was not able to patch pods with completed as true for pods %s in workflow %s due to %s hence abandoning deletion", pod.Name, woc.execWf.ObjectMeta.Name, err)
+			return err
+		}
+	}
+
+	var lselector metav1.LabelSelector
+	if woc.wf.Status.Fulfilled() {
+		log.Info("workflow is fulfilled, deleting everything which may be deleted as per PodGC.Strategy.LabelSelector")
+		lselector = getLabelSelector(wfName, podGC, false)
+	} else {
+		log.Info("deleting existing pods marked as completed")
+		lselector = getLabelSelector(wfName, podGC, true)
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&lselector)
+	if err != nil {
+		log.Errorf("was not able to generate a selector due to %s hence abandoning deletion", err)
+		return nil
+	}
+
+	selectorString := selector.String()
+
+	// optimisation when function is called with a fulfilled workflow
+	// lets patch everything up
+	if woc.wf.Status.Fulfilled() {
+		markDone := true
+		podList, err := pods.List(context.Background(), metav1.ListOptions{LabelSelector: selectorString})
+		if err != nil {
+			log.Errorf("unable to obtain list of pods due to %s", err)
+		} else {
+			for _, nonCompletedPod := range podList.Items {
+				woc.rateLimiter.Wait()
+				_, err := pods.Patch(
```

Comment:
I am probably missing something (that I would like to learn) but if we patch each pod here and then delete them as a batch, what do we gain? Why not just delete them instead? Will we reduce the API calls? Thanks!

Reply:
Thanks for the review @NikeNano! The goal is to reduce the API calls from a series of deletes in rapid succession to batch deletes every so often. We're noticing in some deployments on the major managed Kubernetes providers that workflows that spin up lots of pods that then get deleted in quick succession are being rate limited

There's a bit more context on the issue #10219 but the discussion then shifted to some of the draft PRs. We originally proposed one single batch deletion but @isubasinghe came up with this approach after community members voiced their feedback that they needed to delete pods quicker in order to avoid increased cloud spend

Reply:
@NikeNano we actually do need to patch them, just to state that they have been completed. Argo expects this to be the case for completed workflows(all their pods are marked as completed), so I don't think we should skip this step.

Reply:
Perhaps though in the future to avoid this patching even, we could have a flag that says to the controller to not expect a completed field, this way we can batch delete by just the workflow name.

Reply:
> Argo expects this to be the case for completed workflows(all their pods are marked as completed), so I don't think we should skip this step.

Good point!